Switch to ORM (SQLAchemy)

iodbh opened the pull request:
I've replaced raw SQL statements by SQLAlchemy classes. This should make the code more readable and more straightforward to modify in the future, also allowing us to introduce migrations easily.

This is a pretty big change and I could only test it quickly, so it might have introduced new issues but I think it's a decent enough base to work from.

Please not that database create now works with `python konsent.py createdb` (I haven't updated the README) - drop your existing database for that to work.

surrealbytes commented:
Hey, i can't make this work on python3, have to tested it against python3?
when i try to create the database with `python konsent createdb` i get 
`'Specified key was too long; max key length is 767 bytes'`

iodbh commented:
Yes, I developed/tested on python 3.6.4. This is an SQL error, so my guess is that it is caused in differences in database environment. I can't reproduce it on my box.

I'm using MySQL 5.7.21 (homebrew version) and here's the output of `SHOW TABLE STATUS` ion the konsent db :

```
+----------+--------+---------+------------+------+----------------+-------------+-----------------+--------------+-----------+----------------+---------------------+-------------+------------+-----------------+----------+----------------+---------+
| Name     | Engine | Version | Row_format | Rows | Avg_row_length | Data_length | Max_data_length | Index_length | Data_free | Auto_increment | Create_time         | Update_time | Check_time | Collation       | Checksum | Create_options | Comment |
+----------+--------+---------+------------+------+----------------+-------------+-----------------+--------------+-----------+----------------+---------------------+-------------+------------+-----------------+----------+----------------+---------+
| comments | InnoDB |      10 | Dynamic    |    0 |              0 |       16384 |               0 |        49152 |         0 |              1 | 2018-03-19 10:19:20 | NULL        | NULL       | utf8_general_ci |     NULL |                |         |
| posts    | InnoDB |      10 | Dynamic    |    0 |              0 |       16384 |               0 |        65536 |         0 |              1 | 2018-03-19 10:19:20 | NULL        | NULL       | utf8_general_ci |     NULL |                |         |
| unions   | InnoDB |      10 | Dynamic    |    0 |              0 |       16384 |               0 |        16384 |         0 |              1 | 2018-03-19 10:19:20 | NULL        | NULL       | utf8_general_ci |     NULL |                |         |
| users    | InnoDB |      10 | Dynamic    |    0 |              0 |       16384 |               0 |        32768 |         0 |              1 | 2018-03-19 10:19:20 | NULL        | NULL       | utf8_general_ci |     NULL |                |         |
| votes    | InnoDB |      10 | Dynamic    |    0 |              0 |       16384 |               0 |        65536 |         0 |              1 | 2018-03-19 10:19:20 | NULL        | NULL       | utf8_general_ci |     NULL |                |         |
+----------+--------+---------+------------+------+----------------+-------------+-----------------+--------------+-----------+----------------+---------------------+-------------+------------+-----------------+----------+----------------+---------+
```

iodbh commented:
Look at [This question](https://stackoverflow.com/questions/43411679/cant-create-all-flask-sqlalchemy-specified-key-was-too-long), can you try changing the User's `username` lenght to see if that solves it in your environment ?

surrealbytes commented:
i'll give it a try. it may actually be an issue with MariaDB which is an open source MySQL clone and the default on Debian.

dellitsni commented:
@iodbh Looks neat, good work. There's some stuff we have to fix before merging though:

`createdb` Throws an `OperationalError` in my environment. MariaDB on Windows. I'll give it a try later today using MySQL on Linux. 

Also, Konsent can no longer be executed with `python -m konsent` from the root dir, because it can't find `models.py`. I'm not entirely sure how to work around that, besides putting all files in the same directory, which would make a mess, or adding it to `PATH`, which is pretty clumsy.

iodbh commented:
> Konsent can no longer be executed with python -m konsent from the root dir, because it can't find models.py

`python konsent` from the parent dir works. The way to make the `-m` work would be to use a relative import for models, but that would be part of a reorganization effort. I'm willing to put some work into that once the DB issues have been resolved and this has been merged in order to avoid having an insane branches structure.

Can you provide a full traceback on the `OperationalError` ?

iodbh commented:
Database setups works without any issue on MariaDB on Mac (`Server version: 10.2.13-MariaDB Homebrew`) as well.

dellitsni commented:
My laptop took its last breath today... I won't be working on Konsent for the next week, possibly more, until i get it fixed. I think it's just a broken fan, so I'll be back soon, hopefully. Will still be participating in discussion and so on, I'm glad to see you guys taking interest!

iodbh commented:
Good luck with the laptop ! If you guys can provided more details on your environments (OS and DB versions), I can spin VMs and and get debugging.

surrealbytes commented:
i am on Debian stable. I will test this in the weekend and also work on moving stuff out `__main.py__`

dellitsni commented:
My laptop was on KDE Neon, MySQL i think latest stable version. Can't remember. 

My desktop is running Windows 10, MariaDB 10.1.28.

iodbh commented:
I think I fixed it. Could reproduce the issue on a Debian stable and [this](https://stackoverflow.com/questions/46614894/mariadb-character-encoding) was the source of the issue. @surrealbytes can confirm that my latest commit solves it ?

dellitsni commented:
Still throws an error on Windows 10 with MariaDB, complete stacktrace: https://justpaste.it/1inxt

iodbh commented:
@dellitsni `Unknown database 'konsent'` : did you create the database ?

1. drop the existing database
2. `create database konsent ;`
3. `python konsent createdb`

dellitsni commented:
Oh, i assumed it created the database too. Should have figured that out from the stracktrace, embarassing. Works flawlessly, then. Let's wait for @surrealbytes to test before merging into develop, since he's on Linux.

surrealbytes commented on konsent/__main__.py:
```diff
@@ -1,688 +1,625 @@
+# list unions for pretty-printing
+def print_unions():
+    # find all unions in database
+    unions = Union.qeury.all()
```
typo here, `qeury` instead of `query`

surrealbytes commented on konsent/__main__.py:
```diff
@@ -1,688 +1,625 @@
+            # increment vote value
+            post.votes_count += 1
+            # count that this user has now voted
+            vote = Vote()
```
i get `TypeError: __init__() missing 2 required positional arguments: 'author' and 'target'` here
any idea?

surrealbytes commented:
finally i have time and my PC to test this, the database works perfectly after the name size changes thank you @iodbh! The tests throw some errors, i have commented inline.

iodbh commented:
I've fixed the two above issues and merged. I added an "action" positional arguments, so `python konsent` becomes either `python konsent runserver` or `python konsent createdb`. Short workaround. We can make it cleaner using the `app.cli` (probably #3 )

dellitsni commented:
Nice work, merging!

dellitsni commented:
@iodbh Just noticed another error that i can't quite figure out, you probably know what's going on since you wrote it.

Upvoting a post throws this error: `AttributeError: type object 'Post' has no attribute '_sa_instance_state'`.

iodbh commented:
oops. #29 fixes that.
